File: lib/trace-ir/clock-class.c
/*
 * Copyright 2017-2018 Philippe Proulx <pproulx@efficios.com>
 * Copyright 2013, 2014 Jérémie Galarneau <jeremie.galarneau@efficios.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#define BT_LOG_TAG "CLOCK-CLASS"
#include <babeltrace/lib-logging-internal.h>

#include <babeltrace/assert-pre-internal.h>
#include <babeltrace/compat/uuid-internal.h>
#include <babeltrace/trace-ir/clock-class-const.h>
#include <babeltrace/trace-ir/clock-class.h>
#include <babeltrace/trace-ir/clock-class-internal.h>
#include <babeltrace/trace-ir/clock-snapshot-internal.h>
#include <babeltrace/trace-ir/utils-internal.h>
#include <babeltrace/compiler-internal.h>
#include <babeltrace/types.h>
#include <babeltrace/compat/string-internal.h>
#include <inttypes.h>
#include <babeltrace/object-internal.h>
#include <babeltrace/assert-internal.h>

#define BT_ASSERT_PRE_CLOCK_CLASS_HOT(_cc) \
	BT_ASSERT_PRE_HOT((_cc), "Clock class", ": %!+K", (_cc))

static
void destroy_clock_class(struct bt_object *obj)
{
	struct bt_clock_class *clock_class = (void *) obj;

	BT_LIB_LOGD("Destroying clock class: %!+K", clock_class);

	if (clock_class->name.str) {
		g_string_free(clock_class->name.str, TRUE);
		clock_class->name.str = NULL;
		clock_class->name.value = NULL;
	}

	if (clock_class->description.str) {
		g_string_free(clock_class->description.str, TRUE);
		clock_class->description.str = NULL;
		clock_class->description.value = NULL;
	}

	bt_object_pool_finalize(&clock_class->cs_pool);
	g_free(clock_class);
}

static
void free_clock_snapshot(struct bt_clock_snapshot *clock_snapshot,
		struct bt_clock_class *clock_class)
{
	bt_clock_snapshot_destroy(clock_snapshot);
}

static inline
void set_base_offset(struct bt_clock_class *clock_class)
{
	clock_class->base_offset.overflows = bt_util_get_base_offset_ns(
		clock_class->offset_seconds, clock_class->offset_cycles,
		clock_class->frequency, &clock_class->base_offset.value_ns);
}

struct bt_clock_class *bt_clock_class_create(bt_trace_class *trace_class)
{
	int ret;
	struct bt_clock_class *clock_class = NULL;

	BT_ASSERT_PRE_NON_NULL(trace_class, "Trace class");
	BT_LOGD_STR("Creating default clock class object");

	clock_class = g_new0(struct bt_clock_class, 1);
	if (!clock_class) {
		BT_LOGE_STR("Failed to allocate one clock class.");
		goto error;
	}

	bt_object_init_shared(&clock_class->base, destroy_clock_class);
	clock_class->name.str = g_string_new(NULL);
	if (!clock_class->name.str) {
		BT_LOGE_STR("Failed to allocate a GString.");
		goto error;
	}

	clock_class->description.str = g_string_new(NULL);
	if (!clock_class->description.str) {
		BT_LOGE_STR("Failed to allocate a GString.");
		goto error;
	}

	clock_class->frequency = UINT64_C(1000000000);
	clock_class->origin_is_unix_epoch = BT_TRUE;
	set_base_offset(clock_class);
	ret = bt_object_pool_initialize(&clock_class->cs_pool,
		(bt_object_pool_new_object_func) bt_clock_snapshot_new,
		(bt_object_pool_destroy_object_func)
			free_clock_snapshot,
		clock_class);
	if (ret) {
		BT_LOGE("Failed to initialize clock snapshot pool: ret=%d",
			ret);
		goto error;
	}

	BT_LIB_LOGD("Created clock class object: %!+K", clock_class);
	goto end;

error:
	BT_OBJECT_PUT_REF_AND_RESET(clock_class);

end:
	return clock_class;
}

const char *bt_clock_class_get_name(const struct bt_clock_class *clock_class)
{
	BT_ASSERT_PRE_NON_NULL(clock_class, "Clock class");
	return clock_class->name.value;
}

enum bt_clock_class_status bt_clock_class_set_name(
		struct bt_clock_class *clock_class, const char *name)
{
	BT_ASSERT_PRE_NON_NULL(clock_class, "Clock class");
	BT_ASSERT_PRE_NON_NULL(name, "Name");
	BT_ASSERT_PRE_CLOCK_CLASS_HOT(clock_class);
	g_string_assign(clock_class->name.str, name);
	clock_class->name.value = clock_class->name.str->str;
	BT_LIB_LOGV("Set clock class's name: %!+K", clock_class);
	return BT_CLOCK_CLASS_STATUS_OK;
}

const char *bt_clock_class_get_description(
		const struct bt_clock_class *clock_class)
{
	BT_ASSERT_PRE_NON_NULL(clock_class, "Clock class");
	return clock_class->description.value;
}

enum bt_clock_class_status bt_clock_class_set_description(
		struct bt_clock_class *clock_class, const char *descr)
{
	BT_ASSERT_PRE_NON_NULL(clock_class, "Clock class");
	BT_ASSERT_PRE_NON_NULL(descr, "Description");
	BT_ASSERT_PRE_CLOCK_CLASS_HOT(clock_class);
	g_string_assign(clock_class->description.str, descr);
	clock_class->description.value = clock_class->description.str->str;
	BT_LIB_LOGV("Set clock class's description: %!+K",
		clock_class);
	return BT_CLOCK_CLASS_STATUS_OK;
}

uint64_t bt_clock_class_get_frequency(const struct bt_clock_class *clock_class)
{
	BT_ASSERT_PRE_NON_NULL(clock_class, "Clock class");
	return clock_class->frequency;
}

void bt_clock_class_set_frequency(struct bt_clock_class *clock_class,
		uint64_t frequency)
{
	BT_ASSERT_PRE_NON_NULL(clock_class, "Clock class");
	BT_ASSERT_PRE_CLOCK_CLASS_HOT(clock_class);
	BT_ASSERT_PRE(frequency != UINT64_C(-1) && frequency != 0,
		"Invalid frequency: %![cc-]+K, new-freq=%" PRIu64,
		clock_class, frequency);
	BT_ASSERT_PRE(clock_class->offset_cycles < frequency,
		"Offset (cycles) is greater than clock class's frequency: "
		"%![cc-]+K, new-freq=%" PRIu64, clock_class, frequency);
	clock_class->frequency = frequency;
	set_base_offset(clock_class);
	BT_LIB_LOGV("Set clock class's frequency: %!+K", clock_class);
}

uint64_t bt_clock_class_get_precision(const struct bt_clock_class *clock_class)
{
	BT_ASSERT_PRE_NON_NULL(clock_class, "Clock class");
	return clock_class->precision;
}

void bt_clock_class_set_precision(struct bt_clock_class *clock_class,
		uint64_t precision)
{
	BT_ASSERT_PRE_NON_NULL(clock_class, "Clock class");
	BT_ASSERT_PRE_CLOCK_CLASS_HOT(clock_class);
	BT_ASSERT_PRE(precision != UINT64_C(-1),
		"Invalid precision: %![cc-]+K, new-precision=%" PRIu64,
		clock_class, precision);
	clock_class->precision = precision;
	BT_LIB_LOGV("Set clock class's precision: %!+K", clock_class);
}

void bt_clock_class_get_offset(const struct bt_clock_class *clock_class,
		int64_t *seconds, uint64_t *cycles)
{
	BT_ASSERT_PRE_NON_NULL(clock_class, "Clock class");
	BT_ASSERT_PRE_NON_NULL(seconds, "Seconds (output)");
	BT_ASSERT_PRE_NON_NULL(cycles, "Cycles (output)");
	*seconds = clock_class->offset_seconds;
	*cycles = clock_class->offset_cycles;
}

void bt_clock_class_set_offset(struct bt_clock_class *clock_class,
		int64_t seconds, uint64_t cycles)
{
	BT_ASSERT_PRE_NON_NULL(clock_class, "Clock class");
	BT_ASSERT_PRE_CLOCK_CLASS_HOT(clock_class);
	BT_ASSERT_PRE(cycles < clock_class->frequency,
		"Offset (cycles) is greater than clock class's frequency: "
		"%![cc-]+K, new-offset-cycles=%" PRIu64, clock_class, cycles);
	clock_class->offset_seconds = seconds;
	clock_class->offset_cycles = cycles;
	set_base_offset(clock_class);
	BT_LIB_LOGV("Set clock class's offset: %!+K", clock_class);
}

bt_bool bt_clock_class_origin_is_unix_epoch(const struct bt_clock_class *clock_class)
{
	BT_ASSERT_PRE_NON_NULL(clock_class, "Clock class");
	return (bool) clock_class->origin_is_unix_epoch;
}

void bt_clock_class_set_origin_is_unix_epoch(struct bt_clock_class *clock_class,
		bt_bool origin_is_unix_epoch)
{
	BT_ASSERT_PRE_NON_NULL(clock_class, "Clock class");
	BT_ASSERT_PRE_CLOCK_CLASS_HOT(clock_class);
	clock_class->origin_is_unix_epoch = (bool) origin_is_unix_epoch;
	BT_LIB_LOGV("Set clock class's origin is Unix epoch property: %!+K",
		clock_class);
}

bt_uuid bt_clock_class_get_uuid(const struct bt_clock_class *clock_class)
{
	BT_ASSERT_PRE_NON_NULL(clock_class, "Clock class");
	return clock_class->uuid.value;
}

void bt_clock_class_set_uuid(struct bt_clock_class *clock_class,
		bt_uuid uuid)
{
	BT_ASSERT_PRE_NON_NULL(clock_class, "Clock class");
	BT_ASSERT_PRE_NON_NULL(uuid, "UUID");
	BT_ASSERT_PRE_CLOCK_CLASS_HOT(clock_class);
	memcpy(clock_class->uuid.uuid, uuid, BABELTRACE_UUID_LEN);
	clock_class->uuid.value = clock_class->uuid.uuid;
	BT_LIB_LOGV("Set clock class's UUID: %!+K", clock_class);
}

BT_HIDDEN
void _bt_clock_class_freeze(const struct bt_clock_class *clock_class)
{
	BT_ASSERT(clock_class);

	if (clock_class->frozen) {
		return;
	}

	BT_LIB_LOGD("Freezing clock class: %!+K", clock_class);
	((struct bt_clock_class *) clock_class)->frozen = 1;
}

enum bt_clock_class_status bt_clock_class_cycles_to_ns_from_origin(
		const struct bt_clock_class *clock_class,
		uint64_t cycles, int64_t *ns)
{
	int ret;

	BT_ASSERT_PRE_NON_NULL(clock_class, "Clock class");
	BT_ASSERT_PRE_NON_NULL(ns, "Nanoseconds (output)");
	ret = bt_util_ns_from_origin_clock_class(clock_class, cycles, ns);
	if (ret) {
		ret = BT_CLOCK_CLASS_STATUS_OVERFLOW;
		BT_LIB_LOGW("Cannot convert cycles to nanoseconds "
			"from origin for given clock class: "
			"value overflows the signed 64-bit integer range: "
			"%![cc-]+K, cycles=%" PRIu64,
			clock_class, cycles);
	}

	return ret;
}

void bt_clock_class_get_ref(const struct bt_clock_class *clock_class)
{
	bt_object_get_ref(clock_class);
}

void bt_clock_class_put_ref(const struct bt_clock_class *clock_class)
{
	bt_object_put_ref(clock_class);
}

File: bindings/python/bt2/bt2/clock_class.py
# The MIT License (MIT)
#
# Copyright (c) 2017 Philippe Proulx <pproulx@efficios.com>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

from bt2 import native_bt, object, utils
import uuid as uuidp
import numbers
import bt2
import bt2.clock_snapshot as clock_snapshot


class ClockClassOffset:
    def __init__(self, seconds=0, cycles=0):
        utils._check_int64(seconds)
        utils._check_int64(cycles)
        self._seconds = seconds
        self._cycles = cycles

    @property
    def seconds(self):
        return self._seconds

    @property
    def cycles(self):
        return self._cycles

    def __hash__(self):
        return hash((self.seconds, self.cycles))

    def __eq__(self, other):
        if not isinstance(other, self.__class__):
            # not comparing apples to apples
            return False

        return (self.seconds, self.cycles) == (other.seconds, other.cycles)


class ClockClass(object._SharedObject):
    def __init__(self, name, frequency, description=None, precision=None,
                 offset=None, is_absolute=None, uuid=None):
        utils._check_str(name)
        utils._check_uint64(frequency)
        ptr = native_bt.clock_class_create(name, frequency)

        if ptr is None:
            raise bt2.CreationError('cannot create clock class object')

        super().__init__(ptr)

        if description is not None:
            self.description = description

        if frequency is not None:
            self.frequency = frequency

        if precision is not None:
            self.precision = precision

        if offset is not None:
            self.offset = offset

        if is_absolute is not None:
            self.is_absolute = is_absolute

        if uuid is not None:
            self.uuid = uuid

    def __eq__(self, other):
        if type(self) is not type(other):
            # not comparing apples to apples
            return False

        self_props = (
            self.name,
            self.description,
            self.frequency,
            self.precision,
            self.offset,
            self.is_absolute,
            self.uuid
        )
        other_props = (
            other.name,
            other.description,
            other.frequency,
            other.precision,
            other.offset,
            other.is_absolute,
            other.uuid
        )
        return self_props == other_props

    def __copy__(self):
        return ClockClass(name=self.name, description=self.description,
                          frequency=self.frequency, precision=self.precision,
                          offset=self.offset, is_absolute=self.is_absolute,
                          uuid=self.uuid)

    def __deepcopy__(self, memo):
        cpy = self.__copy__()
        memo[id(self)] = cpy
        return cpy

    def __hash__(self):
        return hash((
            self.name,
            self.description,
            self.frequency,
            self.precision,
            self.offset.seconds,
            self.offset.cycles,
            self.is_absolute,
            self.uuid))

    @property
    def name(self):
        name = native_bt.clock_class_get_name(self._ptr)
        assert(name is not None)
        return name

    @name.setter
    def name(self, name):
        utils._check_str(name)
        ret = native_bt.clock_class_set_name(self._ptr, name)
        utils._handle_ret(ret, "cannot set clock class object's name")

    @property
    def description(self):
        return native_bt.clock_class_get_description(self._ptr)

    @description.setter
    def description(self, description):
        utils._check_str(description)
        ret = native_bt.clock_class_set_description(self._ptr, description)
        utils._handle_ret(ret, "cannot set clock class object's description")

    @property
    def frequency(self):
        frequency = native_bt.clock_class_get_frequency(self._ptr)
        assert(frequency >= 1)
        return frequency

    @frequency.setter
    def frequency(self, frequency):
        utils._check_uint64(frequency)
        ret = native_bt.clock_class_set_frequency(self._ptr, frequency)
        utils._handle_ret(ret, "cannot set clock class object's frequency")

    @property
    def precision(self):
        precision = native_bt.clock_class_get_precision(self._ptr)
        assert(precision >= 0)
        return precision

    @precision.setter
    def precision(self, precision):
        utils._check_uint64(precision)
        ret = native_bt.clock_class_set_precision(self._ptr, precision)
        utils._handle_ret(ret, "cannot set clock class object's precision")

    @property
    def offset(self):
        ret, offset_s = native_bt.clock_class_get_offset_s(self._ptr)
        assert(ret == 0)
        ret, offset_cycles = native_bt.clock_class_get_offset_cycles(self._ptr)
        assert(ret == 0)
        return ClockClassOffset(offset_s, offset_cycles)

    @offset.setter
    def offset(self, offset):
        utils._check_type(offset, ClockClassOffset)
        ret = native_bt.clock_class_set_offset_s(self._ptr, offset.seconds)
        utils._handle_ret(ret, "cannot set clock class object's offset (seconds)")
        ret = native_bt.clock_class_set_offset_cycles(self._ptr, offset.cycles)
        utils._handle_ret(ret, "cannot set clock class object's offset (cycles)")

    @property
    def is_absolute(self):
        is_absolute = native_bt.clock_class_is_absolute(self._ptr)
        assert(is_absolute >= 0)
        return is_absolute > 0

    @is_absolute.setter
    def is_absolute(self, is_absolute):
        utils._check_bool(is_absolute)
        ret = native_bt.clock_class_set_is_absolute(self._ptr, int(is_absolute))
        utils._handle_ret(ret, "cannot set clock class object's absoluteness")

    @property
    def uuid(self):
        uuid_bytes = native_bt.clock_class_get_uuid(self._ptr)

        if uuid_bytes is None:
            return

        return uuidp.UUID(bytes=uuid_bytes)

    @uuid.setter
    def uuid(self, uuid):
        utils._check_type(uuid, uuidp.UUID)
        ret = native_bt.clock_class_set_uuid(self._ptr, uuid.bytes)
        utils._handle_ret(ret, "cannot set clock class object's UUID")

    def __call__(self, cycles):
        return clock_snapshot._ClockSnapshot(self._ptr, cycles)
